File: CSES/Codechef/HailXOR.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

ll powerof2(ll);
void solver(ll [], ll , ll);

ll powerOf2(ll n) {
    ll power = 0;
    while (n) {
        if (n == 1)
            break;
        power++;
        n = n >> 1;
    }
    return power;
}

void solver(ll a[], ll n, ll x) {
    ll i=0, j=i+1;
    while (x > 0) {
        for (;i<n; i++) {
            if (a[i] != 0)
                break;
        }
        for (; j<n; j++) {
            if (a[j] != 0) {
                break;
            }
        }
        ll p = powerOf2(a[i]);
        a[i] = a[i] ^ (1 << p);
        a[j] = a[j] ^ (1 << p);
        x--;
    }
} 

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    ll t;
    cin >> t;
    while (t--) {
        ll n, x;
        cin >> n >> x;
        ll a[n];
        for (ll i=0; i<n; i++) cin >> a[i];
        solver(a, n, x);
        for (int i=0; i<n-1; i++) cout << a[i] << " ";
        cout << a[n-1] << "\n";
    }
}

File: CSES/Codechef/HailXOR.java
package Codechef;

import java.util.Scanner;
class HailXOR {
    static long powerOf2(long n) {
        int power = 0;
        while (n != 0) {
            if (n == 1)
                break;
            n = n >> 1;
            power++;
        }
        return power;
    }

    static void solver(long a[], int x) {
        int i=0;
        while (x-- > 0 && i < a.length) {
            long p = powerOf2(a[i]);
            // System.out.println("p: " + p);
            if ((1 << p) == a[i]) {
                a[i] = 0;
                a[i+1] = a[i+1] ^ (1 << p);
                
                i++;
            }else {
                a[i] = a[i] ^ (1 << p);
                a[i+1] = a[i+1] ^ (1 << p);
            }
            // for (int v:a) System.out.print(v + " ");
            // System.out.println();
        }
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            int n = sc.nextInt();
            int x = sc.nextInt();
            long a[] = new long[n];
            for (int i=0; i<n; i++) a[i] = sc.nextLong();
            solver(a, x);
            for (long v:a) System.out.print(v + " ");
            System.out.println();
        }
        sc.close();
    }
}